sdk/device_update/src/client.rs

```diff
@@ -134,40 +144,57 @@ impl DeviceUpdateClient {
 fn extract_endpoint(url: &Url) -> Result<String> {
     let endpoint = url
         .host_str()
-        .ok_or(Error::DomainParse)?
+        .ok_or_else(|| {
+            Error::with_message(ErrorKind::DataConversion, || {
+                format!("could not get device update domain. url: {url}")
+            })
+        })?
         .splitn(2, '.') // FIXME: replace with split_once() when it is in stable
         .last()
-        .ok_or(Error::DomainParse)?;
+        .ok_or_else(|| {
+            Error::with_message(ErrorKind::DataConversion, || {
+                format!("could not parse device update domain. url: {url}")
+            })
+        })?;
     Ok(format!("{}://{}", url.scheme(), endpoint))
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use azure_core::error::{Result, ResultExt};
 
     #[test]
     fn can_extract_endpoint() -> Result<()> {
+        let url = "https://myadu.api.adu.microsoft.com";
         let suffix = extract_endpoint(
-            &Url::parse("https://myadu.api.adu.microsoft.com")
-                .map_err(|e| CoreError::Http(HttpError::Url(e)))?,
+            &Url::parse(url).with_context(ErrorKind::DataConversion, || {
```

Comment:
These are tests, so context isn't really needed. You can just do `&Url::parse(url)?`. `From` is implemented.
https://github.com/Azure/azure-sdk-for-rust/blob/a6625a5f3b72cc6bf6095c7116d325a79e7816ce/sdk/core/src/error/mod.rs#L237-L241

Reply:
May be check if `url::ParseError` provides the url that it fails on. If it does, then context probably isn't needed in a lot of these.

Reply:
It does not :(, but I think we should not clutter the tests.
I've added one test that should fail though.